### PR TITLE
Add Async QueryData Support

### LINKED ISSUE #116: Avoid running queries on blur
**What happened**:

When writing a query, it can be undesirable for the query to be executed every time the focus is moved from the editor (since it may take a long time for the query to be executed and costs may be incurred).

We are also triggering the function when the table, column or other elements of the editor are changed. This is not necessary either.

**What you expected to happen**:

To have a button to explicitly call the function once the user is done doing modifications.

**How to reproduce it (as minimally and precisely as possible)**:

**Screenshots**

**Anything else we need to know?**:

**Environment**:
- Grafana version:
- Plugin version: 1.0.1
- OS Grafana is installed on:
- User OS & Browser:
- Others:

### LINKED ISSUE #165: Add support for multi-stage queries

### LINKED ISSUE #166: Modify the query editor to add buttons to start and stop a query

### PR DESCRIPTION
**Pull request by iwysiu**
Implements long running queries for athena

Fixes #165
Fixes #166 
Fixes #116 

This PR adds Async Query Data support. This allows queries to be handled over multiple requests instead of a single request. By handling a query over multiple requests (starting, checking its status, and fetching the results) this can help with queries that run for a long time that can potentially timeout. Queries have a unique id that is used to track them between different phases of the query handling flow and this id is useful for preventing queries from being re-run multiple times if it is already running (this can happen if a user refreshes their browser).

To enable, set the `athenaAsyncQueryDataSupport` feature toggle to `true`.

### CODE REVIEW

File: cypress/integration/smoke.spec.ts [hunk 1]
```diff
@@ -74,26 +74,29 @@ e2e.scenario({
...
           },
-          annotations: [
-            {
-              dataSource: 'e2e-athena-datasource',
-              name: 'e2e test annotation',
-              dataSourceForm: () => {
-                e2eSelectors.QueryEditor.CodeEditor.container()
-                  .click({ force: true })
-                  .type(`{selectall} select * from cloudfront_logs where bytes < 100`);
+          // TODO: reenable annotations
```
Comment (kevinwcyu) [inline cypress/integration/smoke.spec.ts:77]: https://github.com/grafana/athena-datasource/issues/187

File: README.md [hunk 2]
```diff
@@ -272,3 +272,9 @@ The backend driver is based on the implementation of [uber/athenadriver](https:/
 - Configure and use [Templates and variables](https://grafana.com/docs/grafana/latest/variables/).
 - Add [Transformations](https://grafana.com/docs/grafana/latest/panels/transformations/).
 - Set up alerting; refer to [Alerts overview](https://grafana.com/docs/grafana/latest/alerting/).
+
+## Async Query Data Support
+
+Async Query Data support enables an asynchronous query handling flow. With Async Query Data support enabled, queries will be handled over multiple requests (starting, checking its status, and fetching the results) instead of having a query be started and resolved over a single request. This is useful for queries that can potentially run for a long time and timeout.
```
Comment (sarahzinger) [inline README.md:278]: might be worth having the docs team check this out if they have time!

File: package.json [hunk 3]
```diff
@@ -19,16 +19,17 @@
...
-    "@grafana/runtime": "8.2.1",
-    "@grafana/toolkit": "8.2.1",
-    "@grafana/ui": "8.2.1",
+    "@grafana/aws-sdk": "0.0.38",
+    "@grafana/data": "8.5.10",
+    "@grafana/e2e": "8.5.10",
+    "@grafana/e2e-selectors": "8.5.10",
+    "@grafana/runtime": "8.5.10",
+    "@grafana/toolkit": "8.5.10",
+    "@grafana/ui": "8.5.10",
```
Comment (sarahzinger) [inline package.json:32]: a thought that I have is that this work could maybe (maybe?) be in multiple prs. Like we upgrade all the packages first for example and release that? Nitpicky but just a suggestion if we're feeling nervous at all!

File: package.json [hunk 4]
```diff
@@ -38,7 +39,7 @@
     "tslib": "^2.3.0"
   },
   "resolutions": {
-    "rxjs": "6.6.3"
+    "rxjs": "^6.6.3"
```
Comment (sarahzinger) [inline package.json:42]: I'm curious if this was necessary or just happened?

File: pkg/athena/api/api.go [hunk 5]
```diff
@@ -73,6 +73,11 @@ func (c *API) Execute(ctx context.Context, input *api.ExecuteQueryInput) (*api.E
 	return &api.ExecuteQueryOutput{ID: *output.QueryExecutionId}, nil
 }
 
+// GetQueryID returns not found, as calling ListQueryExecutions gets throttled when there are many panels
```
Comment (sarahzinger) [inline pkg/athena/api/api.go:76]: Hey sorry I'm not sure I totally understand what this comment/function does, do you mind talking through it a bit more?

File: pkg/athena/driver/connection.go [hunk 6]
```diff
@@ -20,26 +24,53 @@ func newConnection(api *api.API) *conn {
...
+	case athena.QueryExecutionStateQueued:
+		returnStatus = awsds.QuerySubmitted
+	case athena.QueryExecutionStateRunning:
+		returnStatus = awsds.QueryRunning
+	case athena.QueryExecutionStateSucceeded:
+		returnStatus = awsds.QueryFinished
+	case athena.QueryExecutionStateCancelled:
+		returnStatus = awsds.QueryCanceled
+	case athena.QueryExecutionStateFailed:
+		returnStatus = awsds.QueryFailed
```
Comment (sarahzinger) [inline pkg/athena/driver/connection.go:55]: something about this switchcase is so satisfying to me lol. Good work y'all. Maybe worth logging a debug line or something in case people are curious?

File: pkg/athena/driver/driver.go [hunk 7]
```diff
@@ -23,25 +25,29 @@ var (
...
-	return d.connection == nil || d.connection.closed
+	return d.connection == nil || d.asyncDB.closed
 }
 
 func (d *Driver) OpenDB() (*sql.DB, error) {
 	return sql.Open(d.name, "")
 }
 
+func (d *Driver) GetAsyncDB() (awsds.AsyncDB, error) {
+	return d.asyncDB, nil
```
Comment (sarahzinger) [inline pkg/athena/driver/driver.go:46]: feel free to file this under "sarah doesn't get go" but like why do we always return nil if there's no risk of error here? Is the idea that another datasource might have a situation where it needs to return an error so we always write getAsyncDB with the option to return an error?

File: pkg/main.go [hunk 8]
```diff
@@ -5,15 +5,15 @@ import (
...
 	"github.com/grafana/grafana-plugin-sdk-go/backend/datasource"
 	"github.com/grafana/grafana-plugin-sdk-go/backend/log"
-	"github.com/grafana/sqlds/v2"
 )
 
 func main() {
 	// Start listening to requests sent from Grafana.
 	s := athena.New()
-	ds := sqlds.NewDatasource(s)
+	ds := awsds.NewAsyncAWSDatasource(s)
```
Comment (sarahzinger) [inline pkg/main.go:16]: very interesting! I haven't been following along that much with this work maybe at some point y'all could talk about this big refactor and learnings you had along the way? 

I have also run into the question of "is this a sql datasource that uses aws or is this an aws datasource that uses sql" and found the question frustrating... like maybe this is a datasource that has aws-like attributes and sql-like attributes and we want to mixin different util like things. 

None of this is really relevant to this pr of course just talking allow for future work possibilities.

File: src/QueryEditor.test.tsx [hunk 9]
```diff
@@ -95,10 +98,10 @@ describe('QueryEditor', () => {
       ...q,
       connectionArgs: { ...q.connectionArgs, database: 'foo' },
     });
-    expect(onRunQuery).toHaveBeenCalled();
+    expect(onRunQuery).not.toHaveBeenCalled();
```
Comment (sarahzinger) [inline src/QueryEditor.test.tsx:101]: ~remind me why onRunQuery shouldn't run? Do we have like an onRunAsyncQuery or something?~ nevermind I see below we only want to run it now when the run query button is clicked, totally makes sense! I _think_ this could be a separate PR if we wanted, again kind of nitpicking tho.

File: src/datasource.ts [hunk 10]
```diff
@@ -15,7 +16,7 @@ export class DataSource extends DataSourceWithBackend<AthenaQuery, AthenaDataSou
     instanceSettings: DataSourceInstanceSettings<AthenaDataSourceOptions>,
     private readonly templateSrv: TemplateSrv = getTemplateSrv()
   ) {
-    super(instanceSettings);
+    super(instanceSettings, config.featureToggles.athenaAsyncQueryDataSupport);
```
Comment (sarahzinger) [inline src/datasource.ts:19]: so just passing this toggle to super is enough to do all the magic? Wild stuff

**Review by sarahzinger — COMMENTED** [review]
Looking cool! Some questions/suggestions, feel free to push back if they are too much tho! I haven't been super involved in this project so please take any thoughts with a grain of salt. Overall looks really impressive and is going to be much appreciated by our users!!

File: pkg/athena/api/api.go (hunk 5 above)
Comment (kevinwcyu) [reply pkg/athena/api/api.go:76]: This used to use the [`ListQueryExecutions` API](https://docs.aws.amazon.com/athena/latest/APIReference/API_ListQueryExecutions.html) to get a list of query ids and then call [`BatchGetQueryExecutions`](https://docs.aws.amazon.com/athena/latest/APIReference/API_BatchGetQueryExecution.html) to get the current status of a query and to be able to match the currently running query string.

The limit for `ListQueryExecutions` is 5 calls per second https://docs.aws.amazon.com/athena/latest/ug/service-limits.html#service-limits-api-calls so it gets exhausted quite quickly with multiple panels and if we need to page through the results of the call to find the matching query id for the current running query.

File: package.json (hunk 4 above)
Comment (iwysiu) [reply package.json:42]: I think I did that to fix a warning while I was debugging things, but I don't think it's strictly necessary

File: pkg/athena/driver/driver.go (hunk 7 above)
Comment (iwysiu) [reply pkg/athena/driver/driver.go:46]: Yeah, that's more or less it

File: package.json (hunk 4 above)
Comment (iwysiu) [reply package.json]: (undid it to reduce the change)

File: pkg/athena/api/api.go (hunk 5 above)
Comment (iwysiu) [reply pkg/athena/api/api.go]: Rewrote the comment to be clearer.

File: README.md (hunk 2 above)
Comment (kevinwcyu) [reply README.md:278]: @grafana/docs-squad can someone review this if you have the time please?

File: pkg/athena/api/api.go (hunk 5 above)
Comment (sarahzinger) [reply pkg/athena/api/api.go]: rad makes sense thanks!!

File: pkg/athena/driver/driver.go (hunk 7 above)
Comment (sarahzinger) [reply pkg/athena/driver/driver.go:46]: haha cool thanks for bearing with my stupid questions haha

File: src/QueryEditor.test.tsx (hunk 9 above)
Comment (kevinwcyu) [reply src/QueryEditor.test.tsx:101]: will probably just leave this in this pr for now.

File: README.md (hunk 2 above)
Comment (karengermond) [inline README.md:278]:
```suggestion
Async query data support enables an asynchronous query handling flow. With async query data support enabled, queries are handled over multiple requests (starting, checking its status, and fetching the results) instead of starting and resolving a query over a single request. This is useful for queries that can potentially run for a long time and the timeout.
```

File: README.md [hunk 11]
```diff
@@ -272,3 +272,9 @@ The backend driver is based on the implementation of [uber/athenadriver](https:/
...
+
+To enable async query data support, you need to set feature toggle `athenaAsyncQueryDataSupport` to `true`. Here are instructions to [configure feature toggles](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#feature_toggles).
```
Comment (karengermond) [inline README.md:280]:
```suggestion
To enable async query data support, you need to set feature toggle `athenaAsyncQueryDataSupport` to `true`. See [Configure feature toggles](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#feature_toggles) for details.
```
